src/board/arduino/mega2560/parallel.c

```diff
@@ -228,6 +249,55 @@ int parallel_transaction(struct Parallel * port, uint8_t * data, int length, boo
 #define parallel_read(P, D, L) parallel_transaction(P, D, L, true, false)
 #define parallel_write(P, D, L) parallel_transaction(P, D, L, false, false)
 
+// host write -> peripheral read
+// host read -> peripheral write
+int parallel_peripheral_transaction(struct Parallel * port, uint8_t * data, int length, bool read, bool * addr) {
+    if (!read) {
+        // Set data lines as outputs to write to host
+        parallel_data_forward(port);
+    }
+
+    int i;
+    uint8_t byte = 0;
+    for (i = 0; i < length; i++) {
+        // Wait for host to indicate it's written data
+        while (gpio_get(port->data_n) && gpio_get(port->addr_n)) {}
+
+        *addr = !gpio_get(port->addr_n);
+
+        if (read) {
+            data[i] = parallel_read_data(port);
+        } else {
+            byte = data[i];
+            parallel_write_data(port, byte);
+        }
+
+        // Tell host we're processing data
+        gpio_set(port->wait_n, true);
```

Comment:
See http://studenti.fisica.unifi.it/~carla/manuali/porta_parallela/porta_parallela.html

In either cycle, leave the data lines in a floating input state until you know the type of cycle. Wait for either addr_n or data_n to go low `while (addr_n && data_n) {}`. Then read write_n to determine the direction.

In a read cycle (write_n high), set the data lines as outputs and set them to the correct values. Set wait_n high and wait until both addr_n and data_n are high `while (!addr_n || !data_n) {}`. Set data lines as floating inputs and then set wait_n low.

>  An EPP Data Read cycle is shown in Figure 5. The host de-asserts WRITE*, places the AD8-1 lines in a high impedance state, and asserts the DSTROBE* signal. The peripheral then drives the data byte on the AD8-1 lines and re-asserts WAIT* to indicate that the data byte is valid. The host will read the data lines when it sees the WAIT* unasserted, and then de-asserts DSTROBE*. The peripheral then places the AD8-1 lines in an high impedance state and asserts WAIT* to indicate it is ready for the next cycle.

In a write cycle (write_n low), set wait_n high and wait until both addr_n and data_n are high `while (!addr_n || !data_n) {}`. Read the data lines and then set wait_n low.

>  An EPP Data Write cycle is shown in Figure 4. The host first asserts WRITE*, places the data byte on the AD8-1 lines, and asserts DSTROBE*. The peripheral de-asserts WAIT* indicating that it is ready for the data byte. The host then de-asserts DSTROBE* and the peripheral latches the data lines on the rising edge of DSTROBE*. The peripheral then indicates that it is ready for the next cycle by asserting WAIT*.

I recommend using a callback in the peripheral transaction that can perform either address or data, read or write, depending on host request.